src/algorithms/ctmrg.jl

```diff
@@ -1,321 +1,265 @@
-            peps_below_sw = rotate_north(peps_below[rop, col], WEST)
+# Grow environment, compute projectors and renormalize
+function left_move(state, env::CTMRGEnv, alg::CTMRG)
+    corners::typeof(env.corners) = copy(env.corners)
+    edges::typeof(env.edges) = copy(env.edges)
+    ϵ = 0.0
 
-            Q1 = northwest_corner(
-                envs.edges[SOUTH, mod1(row + 1, end), col],
-                envs.corners[SOUTHWEST, mod1(row + 1, end), col],
-                envs.edges[WEST, mod1(row + 1, end), col],
-                peps_above_sw,
-                peps_below_sw,
+    Pleft = similar(state.A, typeof(env.edges[1]))
+    Pright = similar(state.A, typeof(transpose(env.edges[1])))
+    for col in 1:size(state, 2)
+        cnext = _next(col, size(state, 2))
+
+        # Compute projectors
+        for row in 1:size(state, 1)
+            rnext = _next(row, size(state, 1))
+            state_nw = state[row, col]
+            state_sw = rotate_north(state[rnext, col], WEST)
+
+            # Enlarged corners
+            Q_sw = northwest_corner(
+                env.edges[SOUTH, _next(row, end), col],
+                env.corners[SOUTHWEST, _next(row, end), col],
+                env.edges[WEST, _next(row, end), col],
+                state_sw,
             )
-            Q2 = northwest_corner(
-                envs.edges[WEST, row, col],
-                envs.corners[NORTHWEST, row, col],
-                envs.edges[NORTH, row, col],
-                peps_above_nw,
-                peps_below_nw,
+            Q_nw = northwest_corner(
+                env.edges[WEST, row, col],
+                env.corners[NORTHWEST, row, col],
+                env.edges[NORTH, row, col],
+                state_nw,
             )
 
+            # SVD half-infinite environment
             trscheme = if alg.fixedspace == true
-                truncspace(space(envs.edges[WEST, row, cop], 1))
+                truncspace(space(env.edges[WEST, row, cnext], 1))
             else
                 alg.trscheme
             end
-            #@ignore_derivatives @show norm(Q1*Q2)
-
-            (U, S, V) = tsvd(Q1 * Q2; trunc=trscheme, alg=SVD())
-
-            @ignore_derivatives n0 = norm(Q1 * Q2)^2
-            @ignore_derivatives n1 = norm(U * S * V)^2
-            @ignore_derivatives ϵ = max(ϵ, (n0 - n1) / n0)
-
-            isqS = sdiag_inv_sqrt(S)
-            #Q = isqS*U'*Q1;
-            #P = Q2*V'*isqS;
-            @tensor Q[-1; -2 -3 -4] := isqS[-1; 1] * conj(U[2 3 4; 1]) * Q1[2 3 4; -2 -3 -4]
-            @tensor P[-1 -2 -3; -4] := Q2[-1 -2 -3; 1 2 3] * conj(V[4; 1 2 3]) * isqS[4; -4]
+            (U, S, V) = tsvd(Q_sw * Q_nw; trunc=trscheme, alg=TensorKit.SVD())  # TODO: Add field in CTMRG to choose SVD function
+
+            # Compute SVD truncation error and check for degenerate singular values
+            ignore_derivatives() do
+                if unique(x -> round(x; digits=14), diag(S.data)) != diag(S.data) &&
```

Comment:
I definitely agree with factoring out and improving the degeneracy check. I initially coded it up since I needed a  very rudimentary check and so I left it in. I'm happy to adopt your suggestion which does make more sense.